File: sdk/python/shark_auth/audit.py
"""Audit log access — admin API.

Wraps ``/api/v1/audit-logs`` (list/get/export) and
``/api/v1/admin/audit-logs/purge`` (see ``internal/api/router.go``).
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from . import _http
from .proxy_rules import _raise


class AuditClient:
    """Admin client for querying and exporting audit logs.

    Parameters
    ----------
    base_url:
        Base URL of the SharkAuth server.
    admin_api_key:
        Admin API key (``sk_live_...``).
    session:
        Optional pre-configured :class:`requests.Session`.
    """

    _PREFIX = "/api/v1/audit-logs"
    _PURGE = "/api/v1/admin/audit-logs/purge"

    def __init__(
        self,
        base_url: str,
        admin_api_key: str,
        *,
        session: Optional[object] = None,
    ) -> None:
        self._base = base_url.rstrip("/")
        self._token = admin_api_key
        self._session = session or _http.new_session()

    def _auth(self) -> Dict[str, str]:
        return {"Authorization": f"Bearer {self._token}"}

    # ------------------------------------------------------------------
    # List
    # ------------------------------------------------------------------

    def list(
        self,
        *,
        actor_id: Optional[str] = None,
        actor_type: Optional[str] = None,
        action: Optional[str] = None,
        target_id: Optional[str] = None,
        target_type: Optional[str] = None,
        since: Optional[str] = None,
        until: Optional[str] = None,
        limit: int = 100,
        cursor: Optional[str] = None,
    ) -> Dict[str, Any]:
        """List audit log events. GET /api/v1/audit-logs.

        Returns
        -------
        dict
            ``{"events": [...], "next_cursor": ...}`` (shape may vary by
            server version — callers should access fields defensively).
        """
        params: Dict[str, Any] = {"limit": limit}
        if actor_id is not None:
            params["actor_id"] = actor_id
        if actor_type is not None:
            params["actor_type"] = actor_type
        if action is not None:
            params["action"] = action
        if target_id is not None:
            params["target_id"] = target_id
        if target_type is not None:
            params["target_type"] = target_type
        if since is not None:
            params["since"] = since
        if until is not None:
            params["until"] = until
        if cursor is not None:
            params["cursor"] = cursor

        url = f"{self._base}{self._PREFIX}"
        resp = _http.request(
            self._session, "GET", url, headers=self._auth(), params=params
        )
        if resp.status_code == 200:
            body = resp.json()
            if isinstance(body, list):
                return {"events": body, "next_cursor": None}
            if isinstance(body, dict):
                if "events" in body:
                    return body
                if "data" in body and isinstance(body["data"], list):
                    return {
                        "events": body["data"],
                        "next_cursor": body.get("next_cursor"),
                    }
            return body
        _raise(resp)

    # ------------------------------------------------------------------
    # Single event
    # ------------------------------------------------------------------

    def get(self, event_id: str) -> Dict[str, Any]:
        """GET /api/v1/audit-logs/{id}."""
        url = f"{self._base}{self._PREFIX}/{event_id}"
        resp = _http.request(self._session, "GET", url, headers=self._auth())
        if resp.status_code == 200:
            body = resp.json()
            if isinstance(body, dict) and "data" in body and len(body) <= 2:
                return body["data"]
            return body
        _raise(resp)

    # ------------------------------------------------------------------
    # Export
    # ------------------------------------------------------------------

    def export(self, format: str = "ndjson", **filters: Any) -> str:
        """Export audit logs.

        Backend route: ``POST /api/v1/audit-logs/export``. The server
        currently emits CSV; the ``format`` argument is forwarded for
        forward compatibility (``ndjson``/``csv``/``json``).

        ``since`` and ``until`` (ISO 8601) are required by the backend.

        Returns
        -------
        str
            Raw export body text (CSV, NDJSON, or JSON depending on the
            server).
        """
        body: Dict[str, Any] = {"format": format}
        body.update(filters)
        url = f"{self._base}{self._PREFIX}/export"
        resp = _http.request(self._session, "POST", url, headers=self._auth(), json=body)
        if resp.status_code == 200:
            return resp.text
        _raise(resp)

    # ------------------------------------------------------------------
    # Purge
    # ------------------------------------------------------------------

    def purge(
        self,
        *,
        before: Optional[str] = None,
        dry_run: bool = False,
    ) -> Dict[str, Any]:
        """Purge old audit log entries.

        Backend route: ``POST /api/v1/admin/audit-logs/purge``. Deletes
        all entries strictly older than ``before`` (ISO 8601 timestamp).

        Returns the server response body — typically
        ``{"deleted": <int>}`` (or similar).
        """
        body: Dict[str, Any] = {"dry_run": dry_run}
        if before is not None:
            body["before"] = before
        url = f"{self._base}{self._PURGE}"
        resp = _http.request(self._session, "POST", url, headers=self._auth(), json=body)
        if resp.status_code in (200, 202):
            try:
                payload = resp.json()
                if isinstance(payload, dict) and "data" in payload and len(payload) <= 2:
                    return payload["data"]
                return payload
            except Exception:
                return {"raw": resp.text}
        _raise(resp)

File: sdk/typescript/src/audit.ts
/**
 * Audit log access — admin API.
 *
 * Wraps `/api/v1/audit-logs` (list/get/export) and
 * `/api/v1/admin/audit-logs/purge` (see `internal/api/router.go`).
 */

import { httpRequest } from "./http.js";
import { SharkAPIError } from "./errors.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** Audit log event. Open-ended — server shape may vary. */
export interface AuditEvent {
  id?: string;
  action?: string;
  actor_id?: string;
  actor_type?: string;
  target_id?: string;
  target_type?: string;
  status?: string;
  ip?: string;
  created_at?: string;
  metadata?: Record<string, unknown>;
  [key: string]: unknown;
}

/** Filters for {@link AuditClient.list}. */
export interface AuditListFilters {
  actor_id?: string;
  actor_type?: string;
  action?: string;
  target_id?: string;
  target_type?: string;
  /** ISO 8601 lower bound (server may also accept `from`). */
  since?: string;
  /** ISO 8601 upper bound. */
  until?: string;
  limit?: number;
  cursor?: string;
}

/** Result of {@link AuditClient.list}. */
export interface AuditListResult {
  events: AuditEvent[];
  next_cursor?: string | null;
  [key: string]: unknown;
}

/** Filters for {@link AuditClient.export}. */
export interface AuditExportFilters {
  /** ISO 8601 lower bound. Required by backend. */
  since?: string;
  /** ISO 8601 upper bound. Required by backend. */
  until?: string;
  from?: string;
  to?: string;
  action?: string;
  [key: string]: unknown;
}

/** Result of {@link AuditClient.purge}. */
export interface AuditPurgeResult {
  deleted?: number;
  [key: string]: unknown;
}

/** Options for {@link AuditClient}. */
export interface AuditClientOptions {
  baseUrl: string;
  adminKey: string;
}

const PREFIX = "/api/v1/audit-logs";
const PURGE = "/api/v1/admin/audit-logs/purge";
const EXPORT = "/api/v1/audit-logs/export";

// ---------------------------------------------------------------------------
// Client
// ---------------------------------------------------------------------------

/** Admin client for querying and exporting audit logs. */
export class AuditClient {
  private readonly _base: string;
  private readonly _key: string;

  constructor(opts: AuditClientOptions) {
    this._base = opts.baseUrl.replace(/\/+$/, "");
    this._key = opts.adminKey;
  }

  private _auth(): Record<string, string> {
    return { Authorization: `Bearer ${this._key}` };
  }

  private async _throw(status: number, text: string): Promise<never> {
    let code = "api_error";
    let message = text.slice(0, 300);
    try {
      const body = JSON.parse(text) as {
        error?: { code?: string; message?: string } | string;
        message?: string;
      };
      if (typeof body.error === "object" && body.error) {
        if (body.error.code) code = body.error.code;
        if (body.error.message) message = body.error.message;
      } else if (typeof body.error === "string") {
        message = body.error;
      } else if (body.message) {
        message = body.message;
      }
    } catch {
      // keep defaults
    }
    throw new SharkAPIError(message, code, status);
  }

  /** List audit log events. */
  async list(filters: AuditListFilters = {}): Promise<AuditListResult> {
    const qs = new URLSearchParams();
    qs.set("limit", String(filters.limit ?? 100));
    if (filters.actor_id) qs.set("actor_id", filters.actor_id);
    if (filters.actor_type) qs.set("actor_type", filters.actor_type);
    if (filters.action) qs.set("action", filters.action);
    if (filters.target_id) qs.set("target_id", filters.target_id);
    if (filters.target_type) qs.set("target_type", filters.target_type);
    if (filters.since) qs.set("since", filters.since);
    if (filters.until) qs.set("until", filters.until);
    if (filters.cursor) qs.set("cursor", filters.cursor);

    const url = `${this._base}${PREFIX}?${qs.toString()}`;
    const resp = await httpRequest(url, { headers: this._auth() });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);

    const body = resp.json<unknown>();
    if (Array.isArray(body)) {
      return { events: body as AuditEvent[], next_cursor: null };
    }
    if (body && typeof body === "object") {
      const obj = body as Record<string, unknown>;
      if (Array.isArray(obj.events)) {
        return obj as AuditListResult;
      }
      if (Array.isArray(obj.data)) {
        return {
          events: obj.data as AuditEvent[],
          next_cursor: (obj.next_cursor as string | null | undefined) ?? null,
          ...obj,
        };
      }
    }
    return { events: [], next_cursor: null };
  }

  /** Fetch a single audit event by ID. */
  async get(eventId: string): Promise<AuditEvent> {
    const url = `${this._base}${PREFIX}/${encodeURIComponent(eventId)}`;
    const resp = await httpRequest(url, { headers: this._auth() });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    const body = resp.json<AuditEvent | { data: AuditEvent }>();
    if (
      body &&
      typeof body === "object" &&
      "data" in (body as Record<string, unknown>) &&
      Object.keys(body as Record<string, unknown>).length <= 2
    ) {
      return (body as { data: AuditEvent }).data;
    }
    return body as AuditEvent;
  }

  /**
   * Export audit logs.
   *
   * Backend route: `POST /api/v1/audit-logs/export`. Currently emits CSV;
   * `format` is forwarded for forward compatibility.
   *
   * Returns the raw export body text.
   */
  async export(format: string = "ndjson", filters: AuditExportFilters = {}): Promise<string> {
    const body: Record<string, unknown> = { format, ...filters };
    const url = `${this._base}${EXPORT}`;
    const resp = await httpRequest(url, {
      method: "POST",
      headers: { ...this._auth(), "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    return resp.text;
  }

  /**
   * Purge old audit log entries.
   *
   * Backend route: `POST /api/v1/admin/audit-logs/purge`. Deletes entries
   * older than `before` (ISO 8601). Use `dryRun` to count without deleting.
   */
  async purge(before?: string, dryRun: boolean = false): Promise<AuditPurgeResult> {
    const body: Record<string, unknown> = { dry_run: dryRun };
    if (before !== undefined) body.before = before;
    const url = `${this._base}${PURGE}`;
    const resp = await httpRequest(url, {
      method: "POST",
      headers: { ...this._auth(), "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });
    if (resp.status !== 200 && resp.status !== 202)
      await this._throw(resp.status, resp.text);
    try {
      const payload = resp.json<unknown>();
      if (
        payload &&
        typeof payload === "object" &&
        "data" in (payload as Record<string, unknown>) &&
        Object.keys(payload as Record<string, unknown>).length <= 2
      ) {
        return (payload as { data: AuditPurgeResult }).data;
      }
      return payload as AuditPurgeResult;
    } catch {
      return { raw: resp.text } as AuditPurgeResult;
    }
  }
}
